federation/normalize.go

```diff
@@ -0,0 +1,290 @@
+		types[typ] = typ.Name
+
+		for _, field := range typ.Fields {
+			collectTypes(field.Type, types)
+		}
+
+	case *graphql.Union:
+		types[typ] = typ.Name
+		for _, obj := range typ.Types {
+			collectTypes(obj, types)
+		}
+
+	case *graphql.Enum:
+		types[typ] = typ.Type
+
+	case *graphql.Scalar:
+		types[typ] = typ.Type
+
+	default:
+		return fmt.Errorf("bad typ %v", typ)
+	}
+
+	return nil
+}
+
+func makeTypeNameMap(schema *graphql.Schema) (map[string]graphql.Type, error) {
+	allTypes := make(map[graphql.Type]string)
+	if err := collectTypes(schema.Query, allTypes); err != nil {
+		return nil, err
+	}
+	if err := collectTypes(schema.Mutation, allTypes); err != nil {
+		return nil, err
+	}
+
+	reversedTypes := make(map[string]graphql.Type)
+	for typ, name := range allTypes {
+		reversedTypes[name] = typ
+	}
+
+	return reversedTypes, nil
+}
+
+// flattener flattens queries into a normalized form that's easier to wrangle
+// for the query planner and executor.
+//
+// A normalized query has almost all ambiguity removed from the query: Selection
+// sets for objects contain each alias exactly once, and have no fragments.
+// Selection sets for unions (or interfaces) contain exactly one inline fragment
+// with an inner normalized query for each possible type.
+type flattener struct {
+	// types is a map from all type names to the actual type, used to check if a
+	// fragment matches an object type.
+	types map[string]graphql.Type
+}
+
+// complexityLimiter keeps a per-query budget of selections, nodes, etc. to
+// protect against adversarially-constructed queries.
+type complexityLimiter struct {
+	quota int
+}
+
+// allocate reserves n of the complexity budget.
+func (c *complexityLimiter) allocate(n int) error {
+	if c.quota < n {
+		return errors.New("query complexity limit exceeded")
+	}
+	c.quota -= n
+	return nil
+}
+
+// newFlattener creates a new flattener.
+func newFlattener(schema *graphql.Schema) (*flattener, error) {
+	types, err := makeTypeNameMap(schema)
+	if err != nil {
+		return nil, err
+	}
+	return &flattener{
+		types: types,
+	}, nil
+}
+
+// applies checks if obj matches fragment.
+func (f *flattener) applies(obj *graphql.Object, fragment *graphql.Fragment) (bool, error) {
+	switch typ := f.types[fragment.On].(type) {
+	case *graphql.Object:
+		// An object matches if the name matches.
+		return typ.Name == obj.Name, nil
+	case *graphql.Union:
+		// A union matches if the object is part of the union.
+		_, ok := typ.Types[obj.Name]
+		return ok, nil
+	default:
+		return false, fmt.Errorf("unknown fragment type %s", fragment.On)
+	}
+}
+
+// flattenFragments flattens all fragments at the current level. It inlines the
+// selections of each fragment, but does not descend down recursively into those
+// selections.
+func (f *flattener) flattenFragments(selectionSet *graphql.SelectionSet, typ *graphql.Object, c *complexityLimiter, target *[]*graphql.Selection) error {
+	// TODO: is this a good place to limit query complexity?
```

Comment:
How do we want to limit query complexity? Is it saying that there's a limit on how much we can recurse?

Reply:
Yes, something like that, I'm not quite sure. I figured we can return to that later and mostly don't worry about it with query whitelisting.